v/vm.c

```diff
@@ -107,6 +117,57 @@ void printhex(uint64_t x)
   cputstring(str);
 }
 
+// for testing purpose only alloc
+static pte_t alloc()
+{
+  if (freelist_head == 0) {
+    return OUT_OF_MEMORY;
+  }
+  pte_t res = freelist_head->addr;
+  if (freelist_head != freelist_tail)
+    freelist_head = freelist_head->next;
+  else
+    freelist_head = freelist_tail = 0;
+  return res;
+}
+
+static uint64_t map(uint64_t vaddr, uint64_t paddr)
+{
+  // local variables
+  pte_t first_pte = pt[0][VPN2(vaddr)];
+  pte_t* second_pte_ptr;
+  pte_t second_pte;
+  pte_t* third_pte_ptr;
+  pte_t third_pte;
```

Comment:
You don't need separate ptr variables, just use & operator

Reply:
& operator on local variables does not return address to original page table. To modify original pte value, I belive a pointer is necessary.

Reply:
You can go the reverse way if necessary. The point is you don’t need a variable for both the pointer and the pointer contents

Reply:
Sure. Your way is cleaner